@here/harp-mapview/lib/ClipPlanesEvaluator.ts

```diff
@@ -0,0 +1,171 @@
+/*
+ * Copyright (C) 2017-2019 HERE Europe B.V.
+ * Licensed under Apache 2.0, see full license in LICENSE
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { EarthConstants, Projection, ProjectionType } from "@here/harp-geoutils";
+import { assert } from "@here/harp-utils";
+import { Camera } from "three";
+import THREE = require("three");
+
+export interface ClipPlanesEvaluator {
+    /**
+     * Compute near and far clipping planes distance.
+     *
+     * Evaluation method should be called on every frame  and camera clip planes updated.
+     * This is related to evaluator implementation and its input data, that may suddenly change
+     * such as camera position or angle, projection type or so.
+     * Some evaluators may not depend on all or even any of input objects, but to preserve
+     * compatibility with any evaluator type it is strongly reccomended to update on every frame.
+     * @param camera The [[Camera]] object used in scene which clipping planes are evaluated.
+     * @param projection The current [[Projection]] method.
+     * @note Camera clipping planes should be not automatically updated via #evaluateClipPlanes()
+     * call, user should do it manually if needed.
+     */
+    evaluateClipPlanes(camera: Camera, projection: Projection): { near: number; far: number };
+}
+
+/**
+ * Simple camera clip planes evaluator that computes near plance based on ground distance.
+ *
+ * This evaluator supports both planar and spherical projections, although it behaviour is
+ * slightly different in each case. General algorithm setups near plane in between camera and
+ * ground level, while far plane is just
+ */
+export class GroundBasedClipPlanesEvaluator implements ClipPlanesEvaluator {
+    readonly farMin: number;
+
+    protected m_tempVector3: THREE.Vector3 = new THREE.Vector3();
+
+    constructor(
+        readonly nearMin: number = 0.1,
+        readonly nearMultiplier: number = 0.1,
+        readonly nearFarMultiplier = 50.0,
+        readonly farOffset = 200.0
+    ) {
+        assert(nearMin > 0);
+        assert(nearFarMultiplier >= 0);
+        assert(farOffset >= 0);
+        this.farMin = nearMin * nearFarMultiplier + farOffset;
+    }
+
+    evaluateClipPlanes(camera: Camera, projection: Projection): { near: number; far: number } {
+        let nearPlane: number = this.nearMin;
+        let farPlane: number = this.farMin;
+        if (projection.type === ProjectionType.Spherical) {
+            // near and far plane for a set up where
+            // the camera is looking at the center of the scene.
+            const r = EarthConstants.EQUATORIAL_RADIUS;
+            const d = camera.position.length();
+            const alpha = Math.asin(r / d);
+            const xaxis = new THREE.Vector3();
+            const yaxis = new THREE.Vector3();
+            const zaxis = new THREE.Vector3();
+            camera.matrixWorld.extractBasis(xaxis, yaxis, zaxis);
+            const q = new THREE.Quaternion();
```

Comment:
This code is execuded 60 frames per second, why not to store all this vectors and quaternions as members?

Reply:
Good idea, I just tried to leave old behavior (implementation) as much similar as it was, cause this patch-set is providing only new architecture and to simplify comparisons.
Should I change it now?

Reply:
I think, yes. As this PR is for improvement of clipping plane computations, they should be improved as much as possible from this point.